src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,360 @@
+    self.server = None
+    self.setDaemon(True)
+
+  def __del__(self):
+     SasTestHarnessServer.cleanDumpFiles(self.dump_path)
+
+  def getBaseUrl(self):
+    return self.http_server_url
+
+  def getUrlHostNameAndPort(self):
+    """ This method will return the http_server_url used in startServer method by truncating other
+    querystring in url and returns tuple(<hostname>, <port>).
+    """
+    return (self.host_name, self.port)
+
+  def run(self):
+    """ Starts the HTTPServer as background thread.
+    """
+    self.server = SasHttpServer(self.dump_path,
+                                self.getUrlHostNameAndPort(),
+                                SasTestHarnessServerHandler)
+    self.server.socket = ssl.wrap_socket(
+        self.server.socket,
+        certfile=self.cert_file,
+        keyfile=self.key_file,
+        ca_certs=CA_CERT,
+        cert_reqs=ssl.CERT_REQUIRED,
+        ssl_version=ssl.PROTOCOL_TLSv1_2,
+        ciphers=':'.join(CIPHERS),
+        server_side=True
+    )
+
+    # Start HTTP Server.
+    logging.info("Started Test Harness Server:%s at %s", self.name, self.http_server_url)
+    self.server.serve_forever()
+
+  def shutdown(self):
+    self.server.shutdown()
+    logging.info("Stopped Test Harness Server:%s", self.name)
+
+  @staticmethod
+  def getCbsdReferenceId(fcc_id, serial_number):
+    """creates encoded cbsd_reference_id using sha1 with fcc_id and serial number of cbsd.
+    """
+    return str(fcc_id + '/' + str(hashlib.sha1(serial_number).hexdigest())).encode('utf-8')
+
+  @staticmethod
+  def writeDumpFile(base_path, file_name, data):
+    """ Write JSON data to specified file. """
+    dump_file_path = os.path.join(base_path, file_name)
+    file_handler = open(dump_file_path, "w")
+    file_handler.writelines(json.dumps(data, sort_keys=True, indent=4))
+    file_handler.close()
+
+  @staticmethod
+  def cleanDumpFiles(base_path):
+    """ Clean existing dumpfile if any. """
+    map(os.remove, (os.path.join(base_path, file_path)
+                    for file_path in os.listdir(base_path)
+                    if os.path.exists(os.path.join(base_path, file_path))))
+
+  @staticmethod
+  def createFadRecord(url, sas_version, record):
+    """ Creates the files object that includes url,checksum,size,version and recordtype of record
+        required to be packed in FAD.
+    """
+    fad_record = {}
+    fad_record['url'] = url
+    fad_record['checksum'] = hashlib.sha1(json.dumps(record)).hexdigest()
+    fad_record['size'] = sys.getsizeof(record)
+    fad_record['version'] = sas_version
+    if 'cbsd' in url:
+      fad_record['recordType'] = "cbsd"
+    elif 'zone' in url:
+      fad_record['recordType'] = "zone"
+    elif 'esc_sensor' in url:
+      fad_record['recordType'] = "esc_sensor"
+    else:
+      raise Exception("ConfigurationError:: Incorrect Record Type received in URL:{}".format(url))
+    return fad_record
+
+  @staticmethod
+  def createFadObject(all_activity_dump):
```

Comment:
An earlier discussion, with a response from me at the end:

Me:
> Does it make sense to generate the FAD record on the fly? Otherwise you may run into cache invalidation problems.

James:
> Currently the test cases create the files in a local directory and serves those files when requested. These are done in a linear fashion and not simultaneously at moment in the test cases. We made it this based on an earlier comment from that is copied below that mentions that files have to be written locally and then served from there.
> kate-harrison on Jan 9
> Contributor
> In addition to the full dump message itself, we will also need to send back the files containing the CBSDs, PPAs, and ESC sensors. At first glance, this method doesn't appear to support that.
> Edit: I see that you're setting up a separate instance for each piece of data that's included in the SAS TH. I'm not sure this is a sustainable long-term solution and it's a bit hard to read. I thought the solution we discussed at the F2F was to give the SAS TH a local directory from which it would serve files (if there's a local file whose name matches the requested path, it's served; otherwise HTTP 404). Then we can simply swap out files locally and leave the SAS TH running for the duration of the test. Is there a reason you chose to deviate from that?

Reply:

tl;dr: current implementation is OK, no change needed at this time.

I agree that it makes sense to store the CBSD/PPA/ESC files locally and serve them when requested. My suggestion is to auto-create the FAD object (i.e. what is done in this function) when the SAS UUT requests it, not pre-compute ahead of time. This way the user could add/remove CBSD/PPA/ESC files and not have to worry about updating FAD.json, since the TH would scan the dump dir and auto-add the files there to the FAD object.

The current implementation seems to wipe the dir, write the new files, and write the new FAD.json whenever new data is loaded. This has the limitation of requiring us to replace all of the data at once (rather than swap out one of the files), but perhaps this is better in the long run because it forces the code to be more explicit and it should avoid any nasty cache invalidation problems.

Reply:
ok, we can discuss and address this later.